# 01. Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import os

# 02. Importing Dataframes

In [3]:
# Creating path to import dataframes

path = r'/Users/david/Desktop/CF/Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/07-2022 Instacart Basket Analysis'


In [4]:
# Verifying path

path

'/Users/david/Desktop/CF/Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/07-2022 Instacart Basket Analysis'

In [6]:
# Importing 'products.csv' from 'Original Dat' folder as df_prods

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [7]:
# Importing 'orderds_wrangled_2.csv' from 'Prepared Data' folder as 'df_ords'

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled_2.csv'), index_col = False)

FileNotFoundError: [Errno 2] No such file or directory: '/Users/david/Desktop/CF/Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/07-2022 Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled_2.csv'

In [8]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [10]:
# Dropping second index 'Unnamed: 0'

df_ords.drop(columns = ['Unnamed: 0'])

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [11]:
# Redifining 'df_ords'

df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [12]:
df_ords

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

# 03. Consistency Checks

In [9]:
# 1.- Check consistency on 'df_prods' dataframe

# 2.- Run '.describe()' function on 'df_prods'

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

##### There are 49693 observations in the 'df_prods', but the maximum 'product_id' number is 46888, which means there might be 5 observations to look into.

##### The rest of teh variables seem to be okay: 134 different aisles, 21 different departments, and prices look fine (must know the price lists to determine if anything's odd or not)

In [13]:
# 3.- Check for mixed-type data in 'df_ords' dataframe

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
    
    `


In [14]:
# 4.- If any mixed-type data was found, change it to the most convenient data type.

# No mixed-type data was found in any of the variables in 'df_ords'

# Verifying dataframe's variables data types 

df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [15]:
# 5.- Run a check for missing values in 'df_ords' dataframe

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [18]:
# Isolate the 206209 missing values within 'days_since_prior_order' variable/column

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [19]:
df_ords_nan

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [24]:
# checking frequency of values in 'days_since_prior_order' (as confirmation of the null values)

df_ords['days_since_prior_order'].value_counts(dropna = False)


30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [26]:
# Describing 'df_ords', and determine qhich method to use ( Create tags for missing values, inputing mean or median, or dropping values by creating a subset)to handle these missing values

df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [1]:
# 6.- Addressing the missing values

# Considering that the missing values in 'days_since_prior_order' represent 6.02% or the total values listed in the variable,
# I decided to handle these missing values by inputting the mean, as there the biggest frequency value is the one in the extreme of the data distribution, and the mean wouldn't missrepresent the data

df_ords['days_since_prior_order'].fillna(1.111484e+01, inplace=True)

# This function was runned in many other ways, including inputtin the actual value instead of 'value', manually inputting the mean as '1.111484', taking off the 'value' between the parentheses, etc.
# ****Eleanor, would you please tell me what am I doing wrong with this one? Thank you****

NameError: name 'df_ords' is not defined

In [31]:
# As inputting the mean for the missing values did not work out (working with tutor to solve the issue), a subset will be created

df_ords.shape

(3421083, 6)

In [35]:
# Creating a subset of 'df_ords' as 'df_ords_clean' to not include the missing vlaues (these values represent the 6.02% of the values listed within the column 'days_since_prior_order')

df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [36]:
# Verifying 'df_ords_clean'

df_ords_clean

order_id  user_id  order_number  order_day_of_week  \
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
5         3367565        1             6                  2   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
5                        7                    19.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3214874 rows x 6 columns]

In [37]:
df_ords_clean.shape

(3214874, 6)

In [38]:
# The substraction of the 'df_ords_clean' subset/dataframe corresponds to the missing values in 'days_since_prior_order'

3421083-3214874


206209

In [40]:
df_ords_clean['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [43]:
df_ords = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [44]:
df_ords

order_id  user_id  order_number  order_day_of_week  \
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
5         3367565        1             6                  2   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
5                        7                    19.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3214874 rows x 6 columns]

##### The 206209 missing values in 'days_since_prior_order' variable/column represent the 6.02%. The intention was to inputting the mean in the missing values, but since the code didn't work out, I created a subset to then defining the 'df_ords_clean' as 'df_ords'--I decided not to drop the values just to see if I can solver the inputting issue, as I find this way the best as to handle missing values. 

In [46]:
# 7.- Run a check for duplicate values in 'df_ords'

df_dups_ords = df_ords[df_ords.duplicated()]

In [47]:
# Verifying 'df_dups_ords'

df_dups_ords

# 8.- No duplicates found

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [48]:
# 9.- Exporting 'df_ords'

df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked_draft.csv'))

In [12]:
df_ords_cstcy_chk = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', '20_07_22', 'orders_checked_21_07_22.csv'), index_col = False)

In [13]:
df_ords_cstcy_chk_no_dups = df_ords_cstcy_chk.drop_duplicates()

In [14]:
df_ords_cstcy_chk_no_dups['order_id'].value_counts()

2539329    1
1591157    1
1354759    1
1971373    1
1558866    1
          ..
3266950    1
118963     1
9433       1
2938641    1
272231     1
Name: order_id, Length: 3421083, dtype: int64